src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs

```diff
@@ -56,10 +57,18 @@ private IReadOnlyList<RestoreSummaryRequest> GetRequestsFromItems(RestoreArgs re
             SpecValidationUtility.ValidateDependencySpec(dgFile);
 
             // Create requests
-            var requests = new List<RestoreSummaryRequest>();
-            var toolRequests = new List<RestoreSummaryRequest>();
+            var requests = new ConcurrentBag<RestoreSummaryRequest>();
+            var toolRequests = new ConcurrentBag<RestoreSummaryRequest>();
 
-            foreach (var projectNameToRestore in dgFile.Restore)
+            var parallelOptions = new ParallelOptions
+            {
+                // By default, max degree of parallelism is -1 which means no upper bound.
+                // Limiting to processor count reduces task context switching which is better
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+
+            // Parallel.Foreach has an optimization for Arrays, so calling .ToArray() is better and adds almost no overhead
+            Parallel.ForEach(dgFile.Restore.ToArray(), parallelOptions, projectNameToRestore =>
```

Comment:
I have a concern with adding additional scheduled work. 
We blast the threadpool in too many places. 

How does this perform in smaller solutions? Any meaningful improvement there?

Reply:
This change also "clashes" with https://github.com/NuGet/NuGet.Client/pull/3086. 
Once that's merged, I'd expect the gains from this to be lower.

Reply:
This happens before requests are scheduled.  In my profiling, everything happens in parallel except this part of the restore.  The perf gain is much more apparent in large restores, smaller restores will probably not appear to be any faster.

Reply:
In my testing, I disabled the loading of settings per project and still see about .5 to 1 second improvement for 700 projects.  So I think this is still needed along with #3086

Reply:
> This happens before requests are scheduled

That's a good point. :)